beets/library.py

```diff
@@ -715,6 +715,71 @@ def move(self, copy=False, link=False, basedir=None, with_album=True):
 
     # Templating.
 
+    def _legalize_partial(self, path, fragment, replacements):
```

Comment:
Since this is getting rather complex, I propose to break off this method into a pure function so we can test it independently. For this, that would mean passing in `self.path` and the max filename length as arguments; these can then be provided by the `destination` wrapper, which will remain a method.

It might also be useful to give a similar treatment to `_legalize_path` itself. Testability will be really helpful here.